core/ordersync/ordersync.go

```diff
@@ -0,0 +1,507 @@
+				// them toward the number of peers we have successfully synced with.
+				log.WithFields(log.Fields{
+					"provider": peerID.Pretty(),
+				}).Trace("succesfully got orders from peer via ordersync")
+				successfullySyncedPeers.Add(peerID.Pretty())
+			}
+		}
+
+		backoffMut.Lock()
+		delayBeforeNextRetry := retryBackoff.Duration()
+		backoffMut.Unlock()
+		log.WithFields(log.Fields{
+			"delayBeforeNextRetry":    delayBeforeNextRetry.String(),
+			"minPeers":                minPeers,
+			"successfullySyncedPeers": len(successfullySyncedPeers),
+		}).Debug("ordersync could not get orders from enough peers (trying again soon)")
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(delayBeforeNextRetry):
+			continue
+		}
+	}
+
+	return nil
+}
+
+func handleRequestWithSubprotocol(ctx context.Context, subprotocol Subprotocol, requesterID peer.ID, rawReq *rawRequest) (*Response, error) {
+	req, err := parseRequestWithSubprotocol(subprotocol, requesterID, rawReq)
+	if err != nil {
+		return nil, err
+	}
+	return subprotocol.HandleOrderSyncRequest(ctx, req)
+}
+
+func parseRequestWithSubprotocol(subprotocol Subprotocol, requesterID peer.ID, rawReq *rawRequest) (*Request, error) {
+	metadata, err := subprotocol.ParseRequestMetadata(rawReq.Metadata)
+	if err != nil {
+		return nil, err
+	}
+	return &Request{
+		RequesterID: requesterID,
+		Metadata:    metadata,
+	}, nil
+}
+
+func parseResponseWithSubprotocol(subprotocol Subprotocol, providerID peer.ID, rawRes *rawResponse) (*Response, error) {
+	metadata, err := subprotocol.ParseResponseMetadata(rawRes.Metadata)
+	if err != nil {
+		return nil, err
+	}
+	return &Response{
+		ProviderID: providerID,
+		Orders:     rawRes.Orders,
+		Complete:   rawRes.Complete,
+		Metadata:   metadata,
+	}, nil
+}
+
+func (s *Service) getOrdersFromPeer(ctx context.Context, providerID peer.ID) error {
+	stream, err := s.node.NewStream(ctx, providerID, ID)
+	if err != nil {
+		s.handlePeerScoreEvent(providerID, psUnexpectedDisconnect)
+		return err
+	}
+	defer func() {
+		_ = stream.Close()
+	}()
+
+	var nextReq *Request
+	var selectedSubprotocol Subprotocol
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		var rawReq *rawRequest
+		if nextReq == nil {
+			// First request
+			rawReq = &rawRequest{
+				Type:         TypeRequest,
+				Subprotocols: s.SupportedSubprotocols(),
+				Metadata:     nil,
+			}
+		} else {
+			encodedMetadata, err := json.Marshal(nextReq.Metadata)
+			if err != nil {
+				return err
+			}
+			rawReq = &rawRequest{
+				Type:         TypeRequest,
+				Subprotocols: []string{selectedSubprotocol.Name()},
+				Metadata:     encodedMetadata,
+			}
+		}
+
+		if err := json.NewEncoder(stream).Encode(rawReq); err != nil {
+			s.handlePeerScoreEvent(providerID, psUnexpectedDisconnect)
+			return err
+		}
+
+		rawRes, err := waitForResponse(ctx, stream)
+		if err != nil {
+			return err
+		}
+		s.handlePeerScoreEvent(providerID, psValidMessage)
```

Comment:
I see that we parse the response below with the subprotocol, but I still question whether we should give a peer a `psValidMessage` peer score when we cannot decode their response as a valid `rawResponse`.

Reply:
This is addressed by the other changes. We only add the `psValidMessage` score when receiving a valid message.

Reply:
Yep, I just wanted to mention this in case not returning the err was intentional.